### DISCUSSION TITLE
Channelterator hasNext()

### DISCUSSION THREAD

**Discussion by K0zka** [Kotlin/kotlinx.coroutines#3971] [General]
Hi kotlinx.coroutines team,

If you keep calling hasNext() on the channel iterator, it is moving the cursor. So for example
https://pl.kotl.in/5U-Z6RQd1
```
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel


fun main() : Unit = runBlocking {
    val channel = Channel<Int>()
    // producer
    launch {
        channel.send(1)
        channel.send(2)
        channel.close()
    }
    // consumer
    launch {
        val iterator = channel.iterator()
        repeat(10) {
            // only calling hasNext(), nothing else
            println(iterator.hasNext())
        }
    }
}
```

this will generate output 
```
true
true
false
false
...
```

Strictly speaking, this is not the behavior one would expect from a function called **hasNext()** it should be only a query and should not have a side-effect. What do you think?

**Comment by lukellmann** [comment]
I think this will be fixed by https://github.com/Kotlin/kotlinx.coroutines/pull/4065
